vouches/src/lib.rs

```diff
@@ -0,0 +1,105 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Encointer.  If not, see <http://www.gnu.org/licenses/>.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use core::marker::PhantomData;
+use encointer_primitives::vouches::{Vouch, VouchQuality, VouchType};
+use frame_system::{self as frame_system, ensure_signed, pallet_prelude::OriginFor};
+use log::info;
+pub use pallet::*;
+use sp_std::convert::TryInto;
+pub use weights::WeightInfo;
+// Logger target
+const LOG: &str = "encointer::vouches";
+
+#[cfg(feature = "runtime-benchmarks")]
+mod benchmarking;
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+
+mod weights;
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use frame_support::pallet_prelude::*;
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(PhantomData<T>);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config + pallet_timestamp::Config {
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+		type WeightInfo: WeightInfo;
+		#[pallet::constant]
+		type MaxVouchesPerAttester: Get<u32>;
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::call_index(0)]
+		#[pallet::weight((<T as Config>::WeightInfo::vouch_for(), DispatchClass::Normal, Pays::Yes))]
+		pub fn vouch_for(
+			origin: OriginFor<T>,
+			attestee: T::AccountId,
+			vouch_type: VouchType,
+			quality: VouchQuality,
+		) -> DispatchResultWithPostInfo {
+			let attester = ensure_signed(origin)?;
+			let now = <pallet_timestamp::Pallet<T>>::get();
+			let vouch =
+				Vouch { protected: false, timestamp: now, vouch_type, quality: quality.clone() };
+			<Vouches<T>>::try_mutate(
+				&attestee,
+				&attester,
+				|vouches| -> DispatchResultWithPostInfo {
+					vouches.try_push(vouch).map_err(|_| Error::<T>::TooManyVouchesForAttestee)?;
+					Ok(().into())
+				},
+			)?;
+			info!(target: LOG, "vouching: {:?} for {:?}, vouch type: {:?}, quality: {:?}", attester, attestee, vouch_type, quality);
+			Self::deposit_event(Event::VouchedFor { attestee, attester, vouch_type });
+			Ok(().into())
+		}
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// someone or something has vouched for someone or something
+		VouchedFor { attestee: T::AccountId, attester: T::AccountId, vouch_type: VouchType },
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// The calling attester has too many vouches for this attestee
+		TooManyVouchesForAttestee,
+	}
+
+	#[pallet::storage]
+	#[pallet::getter(fn vouches)]
+	pub(super) type Vouches<T: Config> = StorageDoubleMap<
+		_,
+		Identity,
+		T::AccountId,
+		Identity,
+		T::AccountId,
+		BoundedVec<Vouch<T::Moment>, T::MaxVouchesPerAttester>,
```

Comment:
Here I am just wondering. Is this going to be a significant number to just have a theoretical upper bound for the PoV, or do we indeed want to limit the number of vouches per attester to something like 50? Probably the former to support arbitrary use cases, but I think the latter could be interesting as well. :)

Reply:
Yes, this is meant to bound the PoV size.
I envision to start with a limit of 32.